src/sync/deduplicator.ts

```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @param jobKey - a key of the job to be used for "deduplication", e.g. installationId
+	 * @param executionId
+	 */
+	setInProgressFlag: (jobKey: string, executionId: string) => Promise<void>;
+
+	/**
+	 * Should be called when the executor finishes its execution of the job
+	 *
+	 * @param jobKey - a key of the job to be used for "deduplication", e.g. installationId
+	 */
+	removeInProgressFlag: (jobKey: string) => Promise<void>;
+
+	/**
+	 * Should be called by executor to figure out if there's currently already a job running or not. Yes - check
+	 * the status of it (see isExecutionLive), no - feel free to run it.
+	 *
+	 * @param jobKey - a key of the job to be used for "deduplication", e.g. installationId
+	 * @param invalidatingTimestamp - time after which the flag is considered stale and can be ignored
+	 * 											(a use-case is when a node dies during the execution of the job)
+	 * @return previously set executionId if the flag is raised
+	 */
+	hasInProgressFlag: (jobKey: string, invalidatingTimestamp: number) => Promise<string | null>
+
+	/**
+	 *
+	 * @param jobKey - a key of the job to be used for "deduplication", e.g. installationId
+	 * @param executionId - the ID of the execution to watch for
+	 * @param executorFlagUpdateTimeoutMsecs - the timeout which executor uses to update the flag with, the value is
+	 * 																				 used to calculate the sleep interval between observations of the flag
+   *																				 during the execution.
+	 *
+	 * @return true if and only if can 100% tell that the execution with the given ID  is during something (the process is
+	 * live), false otherwise (it might be still live but we couldn't determine that)
+	 */
+	isExecutionLive: (jobKey: string, executionId: string, executorFlagUpdateTimeoutMsecs: number) => Promise<boolean>;
+}
+
+function sleep(ms) {
+	return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+type Flag = {
+	executionId: string,
+	timestamp: number
+};
+
+export class RedisInProgressStorageWithTimeout implements InProgressStorage {
+	private redis: Redis;
+
+	constructor(redis: Redis) {
+		this.redis = redis;
+	}
+
+	hasInProgressFlag(jobKey: string, invalidatingTimestamp: number): Promise<string | null> {
+		return this.redis.get(jobKey).then(json => {
+			if (!json) {
+				return null;
+			}
+			const flag = JSON.parse(json) as Flag;
+			const isStaled = (new Date().getTime() - flag.timestamp) >= invalidatingTimestamp;
+			if (isStaled) {
+				return null;
+			}
+			return flag.executionId;
+		});
+	}
+
+	async removeInProgressFlag(jobKey: string): Promise<void> {
+		await this.redis.unlink(jobKey);
+	}
+
+	async setInProgressFlag(jobKey: string, executionId: string): Promise<void> {
+		const flag: Flag = {
+			executionId,
+			timestamp: new Date().getTime()
+		};
+		// We don't want to pollute redis, autoexpire after the flag is not being updated
+		const REDIS_CLEANUP_TIMEOUT = 24 * 3600 * 1000;
```

Comment:
24 hours looks like a huge TTL, especially when job updates it periodically

Reply:
This TTL is for garbage collection; it is not an operational one.

Reply:
Didn't know! Thanks for clarifying